src/main/java/org/tikv/txn/TwoPhaseCommitter.java

```diff
@@ -0,0 +1,508 @@
+    private static final int bytesPerMiB = 1024 * 1024;
+    private long maxTxnTimeUse = 60_1000;//unit is: milliseconds
+    // ttl = ttlFactor * sqrt(writeSizeInMiB)
+    private static final int ttlFactor = 6000;
+    private static final int maxLockTTL = 12000;
+    private static final SecondaryCommitTaskThreadPool secondaryCommitPool = new SecondaryCommitTaskThreadPool();
+
+    private Map<String, Kvrpcpb.Mutation> mutations = new LinkedHashMap<>();
+    private List<byte[]> keysList;
+    //private ReadOnlyPDClient pdClient;
+    private TxnKVClient kvClient;
+
+    private RegionManager regionManager;
+
+    private long lockTTL = 0;
+    /**
+     * start timestamp of transaction which get from PD
+     */
+    private long startTs = 0;
+    /**
+     * commit timestamp of transaction which get from PD
+     */
+    private long commitTs = 0;
+
+    private volatile boolean prewriteTaskError = false;
+    //private volatile AtomicInteger seondaryThreadIdGenerator = new AtomicInteger(0);
+
+    private TwoPhaseCommitter() {}
+
+    public TwoPhaseCommitter(ITransaction transaction) {
+        //this.pdClient = transaction.getKVClient().getSession().getPDClient();
+        this.regionManager = transaction.getKVClient().getRegionManager();
+        this.keysList = new LinkedList<>();
+        this.kvClient = transaction.getKVClient();
+        this.startTs = transaction.getStartTS();
+
+        Map<byte[], byte[]> storedKeys = transaction.getStoredKeys();
+        int putCount = 0, delCount = 0, lockCount = 0;
+        int txnSize = 0;
+        for(byte[] key : storedKeys.keySet()) {
+            byte[] value = storedKeys.get(key);
+            if(value.length > 0) {
+                Kvrpcpb.Mutation mutation = Kvrpcpb.Mutation.newBuilder()
+                        .setKey(ByteString.copyFrom(key))
+                        .setValue(ByteString.copyFrom(value))
+                        .setOp(Kvrpcpb.Op.Put)
+                        .build();
+                mutations.put(new String(key), mutation);
+                putCount++;
+            } else {
+                Kvrpcpb.Mutation mutation = Kvrpcpb.Mutation.newBuilder()
+                        .setKey(ByteString.copyFrom(key))
+                        .setOp(Kvrpcpb.Op.Del)
+                        .build();
+                mutations.put(new String(key), mutation);
+                delCount ++;
+            }
+            keysList.add(key);
+            txnSize += (key.length + value.length);
+            //TODO check transaction maxEntrySize
+        }
+        List<byte[]> lockedKeys = transaction.getLockedKeys();
+        for(byte[] lockedKey : lockedKeys) {
+            Kvrpcpb.Mutation mutation = Kvrpcpb.Mutation.newBuilder()
+                    .setKey(ByteString.copyFrom(lockedKey))
+                    .setOp(Kvrpcpb.Op.Lock)
+                    .build();
+            mutations.put(new String(lockedKey), mutation);
```

Comment:
If a key is already contained in `mutations`, we should not use the lock mutation to replace the old one.

Reply:
```suggestion
            ByteString key = ByteString.copyFrom(lockedKey);
            if(this.mutations.get(key) != null) {
                //locked key exist already, omit it.
                continue;
            }
            Kvrpcpb.Mutation mutation = Kvrpcpb.Mutation.newBuilder()
                    .setKey(key)
                    .setOp(Kvrpcpb.Op.Lock)
                    .build();
            mutations.put(new String(lockedKey), mutation);
```